### DISCUSSION TITLE
Modifying/parameterize a column mapping at runtime.

### DISCUSSION THREAD

**Discussion by jenstroeger** [sqlalchemy/sqlalchemy#10705] [Usage Questions] [answered]
Hi,

Suppose I have a [custom type](https://docs.sqlalchemy.org/en/20/core/custom_types.html) like so:
```python
class Encrypted(TypeDecorator):
    impl = LargeBinary
    cache_ok = True

    def __init__(self, cipher = None) -> None:
        super().__init__()
        self.cipher = cipher  # Allowing None makes migrations and defaults easier.

    def process_bind_param(self, value: str | None, dialect: Dialect) -> bytes | None:
        if value is None:
            return None
        if self.cipher is None:
            raise ValueError("Unable to encrypt data without cipher")
        return cipher.encrypt(value)

    def process_result_value(self, value: bytes | None, dialect: Dialect) -> str | None:
        if value is None:
            return None
        if self.cipher is None:
            raise ValueError("Unable to decrypt data without cipher")
        return cipher.decrypt(value)

    @property
    def python_type(self) -> type:
        return str
```
Using this type I can create a [declarative mapping](https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#declarative-mapping):
```python
class Person(Base):
    __tablename__ = "person"
    super_secret_data: str | None = Column(Encrypted(), nullable=True)
```
This mapping works for as long as I don’t access the mapped `super_secret_data` property. I need a cipher to be able to en/decrypt that data. However, those ciphers need to be changed around for different `Person` instances.

I’ve tried to modify the mapping’s cipher in place but that didn’t seem to have an effect:
```python
Person.__table__.columns["super_secret_data"].type.cipher = new_cipher
```

The current solution takes the [imperative mapping](https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#imperative-mapping) approach:
```python
CipheredPerson = type(f"Encrypted{id(cipher)}", (), {})
Base.registry.map_imperatively(
    mapping_type,
    Table(
        "person",
        Base.metadata,
        Column(super_secret_data, Encrypted(cipher=cipher), nullable=True),
        extend_existing=True,  # Because we already have the `Person` mapping, and likely other `CipheredPerson` mappings.
    ),
)
person = CipheredPerson()  # Person mapping with cipher.
person.super_secret_data = "this is a secret!"
```
So for every cipher I create a new mapping and access the underlying table’s `super_secret_data` column through that mapping. The mappings are uniqued using the cipher’s [identity](https://docs.python.org/3/library/functions.html#id) (not perfect, but works for now).

This feels clunky and weird to me, though.

My question: is there a better way to modify the column mapping’s cipher? How does this approach scale (e.g. thousands of dynamically created mappings)?

**Comment by zzzeek** [comment]
> However, those ciphers need to be changed around for different Person instances.

OK, then you can't change state on a TypeEngine object because that is not per instance...

> I’ve tried to modify the mapping’s cipher in place but that didn’t seem to have an effect:

that's changing the type for `Person` globally....

> The current solution takes the [imperative mapping](https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#imperative-mapping) approach:

that looks like you are creating a new mapping...for every **instance** of `Person`, so yes that is very wrong.

if you are doing per-instance `Person` stuff then you cannot do the encrypt/decrypt within the `TypeEngine` object, you have to do it at the level of the `Person` class and I would do this using an ordinary [Python descriptor](https://docs.python.org/3/howto/descriptor.html).


```py

class Person(Base):
    # ...
    
    encrypted_content: Mapped[str]

    @property
    def decrypted_content(self):
        return decrypt(self.encrypted_content, self.cipher)

    @decrypted_content.setter
    def decrypted_content(self, value):
        self.encrypted_content = encrypt(value, self.cipher)
```

**Reply by jenstroeger** [reply]
Thanks @zzzeek, and I suspect creating different mappings on the fly with different ciphers is just cumbersome by design…

Your suggestion, though, hoists the “en/decrypt” responsibility from the column mapping alone to the entirety of the row mapping. Might there be a way to keep _only_ the column mapping aware of the cipher, so that it can be reused for different table mappings?

> that looks like you are creating a new mapping...for every instance of `Person`, so yes that is very wrong.

What’s the impact of that on SQLA internals? I’m curious to learn a bit more here 🤓

**Comment by zzzeek** [comment]
> Thanks @zzzeek, and I suspect creating different mappings on the fly with different ciphers is just cumbersome by design…
> 
> Your suggestion, though, hoists the “en/decrypt” responsibility from the column mapping alone to the entirety of the row mapping. Might there be a way to keep _only_ the column mapping aware of the cipher, so that it can be reused for different table mappings?

this would be best achieved using a mixin class like HasEncryptedThing


```py

class HasEncryptedThing:
    # ...
    
    encrypted_content: Mapped[str]

    @property
    def decrypted_content(self):
        return decrypt(self.encrypted_content, self.cipher)

    @decrypted_content.setter
    def decrypted_content(self, value):
        self.encrypted_content = encrypt(value, self.cipher)

class Person(HasEncryptedThing, Base):
    # ...

```

within SQLAlchemy we are trying to make a reasonable design possible but we aren't getting to excessive amounts of magic where we can avoid it.


> What’s the impact of that on SQLA internals? I’m curious to learn a bit more here 🤓

well lots of classes and mapped classes, depending on how many distinct Person classes you are making, mappers are expensive to create, are not well suited to on-the-fly creation, cache keys from queries would be totally unique to each mapper so you lose the benefits of caching, subclasses of Person become not reasonable, etc.   this is really about instance-level vs. class-level behavior it's a standard OOP problem.

**Reply by jenstroeger** [reply]
> this would be best achieved using a mixin class like HasEncryptedThing

Thanks @zzzeek. For fixed column names or for one single column per table a mixin works. In my case, though, I need to be flexible with the column names and the number of columns per table.

What I cooked up for now is a function that creates the mixin classes:
```python
def create_encrypted_mixin(column_name: str) -> type:
    @declarative_mixin
    class EncryptedMixin:
        pass

    setattr(EncryptedMixin, column_name, Column(column_name, sa.LargeBinary, nullable=True))
    setattr(EncryptedMixin, f"encrypt_{column_name}", lambda self, cipher: None if (value := getattr(self, column_name)) is None else decrypt(cipher, value))
    setattr(EncryptedMixin, f"decrypt_{column_name}", lambda self, cipher: None if (value := getattr(self, column_name)) is None else decrypt(cipher, value))

    return EncryptedMixin

ThisEncryptedMixin = create_encrypted_mixin("this_value")
ThatEncryptedMixin = create_encrypted_mixin("that_value")
```
I thought about opening another question and ask if it’s at all possible to use mixins with variable column names.

What leaves me uneasy about this implementation are pylint (e.g. [duplicate-bases](https://pylint.pycqa.org/en/latest/user_guide/messages/error/duplicate-bases.html)[^1]) and mypy errors (e.g. [variable not valid as a type](https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases)), both of which indicating that this approach is a bit crooked.

An alternative to the above would be a simple base class from which we then derive the various mixin classes, but that would leave more work (and responsibility) with the user. And maybe that’s just fine because it’s clean and simple and closer to the Python mantra… 🤔

[^1]: That, I think, could be addressed by creating a single mixin class for _all_ columns, but then I can’t create multiple of those classes. Or I create different type names in the function, but then we’re back to the initial question whether that scales.

**Reply by zzzeek** [reply, marked as answer]
So that new API you have there seems to have changed, where now you want to pass the cipher with the value itself, rather than having it be part of the class.

if you are doing *that*, then you can actually go back to your `TypeDecorator` version and just pass your own `Encrypted` object  to it.  why not?

```py
class Encrypted:
     def __init__(self, content, cipher):
         self.encrypted = encrypt(content, cipher)

    def decrypt(self, cipher):
        return decrypt(self.encrypted, cipher)

# ...


p1 = Person(encrypted_content = Encrypted("the content", cipher))

print(p1.encrypted_content.decrypt(cipher))
```

**Reply by jenstroeger** [reply]
> So that new API you have there seems to have changed, where now you want to pass the cipher with the value itself, rather than having it be part of the class.

It’s another way of achieving one goal — different rows (mapped objects) need to have their own cipher.

The initial attempt to solve that was to create a custom type on the fly and pass to the cipher to that custom type. The second attempt then was to use a mixin that provided a function that takes a cipher and decrypts a particular column. In both cases I could use different ciphers on the mapped objects.

> if you are doing _that_, then you can actually go back to your `TypeDecorator` version and just pass your own `Encrypted` object to it. why not?

The `Encrypted` in your example is _different_ than the `Encrypted(TypeDecorator)` custom type in my initial example? I’m a bit confused now 🤔

But I think I understand what you’re saying: the `Encrypted(TypeDecorator)` should take a “wrapper object” (which you also called `Encrypted` in the example) that handles the per-row cipher, and encrypts as it passes the data _to_ the `Encrypted` column and it decrypts as it reads _from_ the `Encrypted` column. From the fixed custom type perspective it’s just bytes flowing in and out then. Is that what you suggest? (It’s an approach similar to what [sqlalchemy-utils](https://github.com/kvesteri/sqlalchemy-utils) is doing with e.g. the [CurrencyType](https://sqlalchemy-utils.readthedocs.io/en/latest/data_types.html#module-sqlalchemy_utils.types.currency) that requires a [Currency](https://sqlalchemy-utils.readthedocs.io/en/latest/data_types.html#sqlalchemy_utils.primitives.currency.Currency) object.)

**Reply by zzzeek** [reply]
> 
> But I think I understand what you’re saying: the `Encrypted(TypeDecorator)` should take a “wrapper object” (which you also called `Encrypted` in the example) that handles the per-row cipher, and encrypts as it passes the data _to_ the `Encrypted` column and it decrypts as it reads _from_ the `Encrypted` column. From the fixed custom type perspective it’s just bytes flowing in and out then. Is that what you suggest? 

that's the suggestion, yes

**Reply by jenstroeger** [reply]
> that's the suggestion, yes

Yup that worked, and the devs just have to work with the `EncryptedValue` wrapper class for the `Encrypted` custom type 👍🏼  I’ll accept that suggestion as the answer.
